keps/sig-scheduling/20200116-coscheduling-plugin-based-on-scheduler-framework.md

```diff
@@ -0,0 +1,222 @@
+    ...
+     // Count the numbers of all pods belongs to the same PodGroup excludes terminating ones.
+    total := cs.calculateTotalPods(podGroupName)
+    if total < minAvailable {
+        return framework.NewStatus(framework.UnschedulableAndUnresolvable, "less than minAvailable")
+    }
+    ...
+    if cs.Last != "" && cs.Last != podGroupName {
+        allocated := cs.calculateAllocatedPods(cs.LastPodGroup)
+        pgInfo, ok := cs.PodGroupInfos[cs.LastPodGroup]
+        if ok {
+        	pgInfo.LastFailureTimestamp = time.Now()
+        	if allocated < pgInfo.minAvailable {
+            	cs.FrameworkHandle.IterateOverWaitingPods(func(p framework.WaitingPod) {
+                	if p.GetPod().Labels[PodGroupName] == cs.LastPodGroup {
+                    	p.Reject(cs.Name())
+                	}
+            	})
+        	}
+        }
+    }
+    cs.LastPodGroup = podGroupName
+    return nil
+}
+```
+
+#### Permit
+In `Permit` phase, we put the pod that doesn't meet min-available into the WaitingMap and reserve resources until min-available are met or timeout is triggered.
+1. Get the number of Running pods that belong to the same PodGroup
+2. Get the number of WaitingPods (used to record pods in waiting status) that belong to the same PodGroup
+3. If Running + WaitingPods + 1 >= min-available(1 means the pod itself), approve the waiting pods that  belong to the same PodGroup. Otherwise, put the pod into WaitingPods and set the timeout (eg: 10s).
+ 
+```go
+func (cs Coscheduling) Permit(ctx context.Context, state *framework.CycleState, p *v1.Pod, nodeName string) (*framework.Status, time.Duration) {
+    ...
+    running := cs.calculateRunningPods(podGroupName)
+    waiting := cs.calculateWaitingPods(podGroupName)
+    if running+waiting+1 < minAvailable {
+        return framework.NewStatus(framework.Wait, ""), 10 * time.Second
+    }
+    ...
+    cs.FrameworkHandle.IterateOverWaitingPods(func(p framework.WaitingPod) {
+        if p.GetPod().Labels[PodGroupName] == podGroupName {
+        	p.Allow(cs.Name())
+        }
+    })
+   
+    delete(cs.PodGroupInfos, podGroupName)
+    return framework.NewStatus(framework.Success, ""), 0
+}
+```
+ 
+
+#### UnReserve
+After a pod which belongs to a PodGroup times out in the permit phase.  UnReserve ```Rejects``` the pods that belong to the same PodGroup to avoid long-term invalid reservation of resources.
+
+```go 
+func (cs *Coscheduling) Unreserve(ctx context.Context, state *framework.CycleState, p *v1.Pod, nodeName string) {
+    ...
+    cs.FrameworkHandle.IterateOverWaitingPods(func(p framework.WaitingPod) {
+        if p.GetPod().Labels[PodGroupName] == podGroupName {
+        	p.Reject(cs.Name())
+        }
+    })
+}
+```
+
+## Alternatives considered
+1. Using `PodGroup` as a scheduling unit. This requires major refactoring, which only supports Pods as scheduling unit today.
+
+
```

Comment:
`Risks and Mitigations `are also needed, for example the current scheduler's performance is at 100/s. If `PodGroup` is enabled, then some measures are needed to ease the pressure on the scheduler.